packages/medusa/src/services/product.ts

```diff
@@ -449,12 +478,14 @@ class ProductService extends BaseService {
     return result.id
   }
 
-  async upsertProductTags_(tags) {
+  private async upsertProductTags_(
+    tags: CreateProductProductTagDTO[]
+  ): Promise<ProductTag[]> {
     const productTagRepository = this.manager_.getCustomRepository(
       this.productTagRepository_
     )
 
-    const newTags = []
+    const newTags: ProductTag[] = []
     for (const tag of tags) {
       const existing = await productTagRepository.findOne({
```

Comment:
I like it! I went ahead and completely moved the methods to the repo layer where I believe they fit better. I removed these methods from the `ProductService` and just replaced wherever they were consumed by the method from the appropriate repositories

Reply:
I see the reason to put the part for upsert in the repo, but I think it is overkill to move the map construction into a separate method :) it gives some context to have it among the little rest.
Here is what I think it could looks like:
```ts
 public async upsertTags(tags: UpsertTagsInput): Promise<ProductTag[]> {
    const tagsValues = tags.map((tag) => tag.value)
    const existingTags = await this.find({
      where: {
        value: In(tagsValues),
      },
    })
    const existingTagsMap = new Map<string, ProductTag>(
      existingTags.map((tag) => [tag.value, tag])
    )

    const upsertedTags: ProductTag[] = []

    for (const tag of tags) {
      const tagExists = existingTagsMap.get(tag.value)
      if (tagExists) {
        upsertedTags.push(aTag)
        continue
      }
    
      const newTag = this.create(tag)
      const savedTag = await this.save(newTag)
      upsertedTags.push(savedTag)
    }

    return upsertedTags
}
```